thali/thalireplicationmanager.js

```diff
@@ -0,0 +1,301 @@
+*/
+ThaliReplicationManager.prototype.start = function (deviceName, port, dbName) {
+  validations.ensureNonNullOrEmptyString(deviceName, 'deviceName');
+  validations.ensureValidPort(port);
+  validations.ensureNonNullOrEmptyString(dbName, 'dbName');
+
+  this.emit(ThaliReplicationManager.events.STARTING);
+
+  this._port = port;
+  this._deviceName = deviceName;
+  this._dbName = dbName;
+  this._serverBridge = muxServerBridge.call(this, port);
+  this._serverBridge.listen(function () {
+    this._serverBridgePort = this._serverBridge.address().port;
+
+    this._emitter.startBroadcasting(deviceName, this._serverBridgePort, function (err) {
+      if (err) {
+        this._isStarted = false;
+        this.emit(ThaliReplicationManager.events.START_ERROR, err);
+      } else {
+        this._isStarted = true;
+        this._emitter.addListener(PEER_AVAILABILITY_CHANGED, syncPeers.bind(this));
+        this._emitter.addListener(NETWORK_CHANGED, networkChanged.bind(this));
+        this.emit(ThaliReplicationManager.events.STARTED);
+      }
+    }.bind(this));
+  }.bind(this));
+};
+
+/**
+* Starts the Thali replication manager with the given device name and port number
+* @param {String} deviceName the device name to advertise.
+* @param {Number} port the port number used for synchronization.
+*/
+ThaliReplicationManager.prototype._restart = function () {
+  this.emit(ThaliReplicationManager.events.STARTING);
+
+  this._serverBridge = muxServerBridge.call(this, this._port);
+  this._serverBridge.listen(function () {
+    this._serverBridgePort = this._serverBridge.address().port;
+    this.emit(ThaliReplicationManager.events.STARTED);
+  }.bind(this));
+};
+
+/**
+* Stops the Thali replication manager
+*/
+ThaliReplicationManager.prototype.stop = function () {
+  if (!this._isStarted) { throw new Error('.start must be called before stop'); }
+  this.emit(ThaliReplicationManager.events.STOPPING);
+
+  this._emitter.stopBroadcasting(function (err) {
+    if (err) {
+      this.emit(ThaliReplicationManager.events.STOP_ERROR, err);
+    } else {
+      this._emitter.removeAllListeners(PEER_AVAILABILITY_CHANGED);
+      this._emitter.removeAllListeners(NETWORK_CHANGED);
+
+      Object.keys(this._replications).forEach(function (key) {
+        var item = this._replications[key];
+        item.from.cancel();
+        item.to.cancel();
+      }, this);
+
+      this._serverBridge.close();
+      this._isStarted = false;
+      this.emit(ThaliReplicationManager.events.STOPPED);
+    }
+  });
+};
+
+function networkChanged(status) {
+  if (!status.isAvailable && this._isStarted) {
+    this.stop();
+  }
+
+  if (status.isAvailable && !this._isStarted) {
+    this.start(this._deviceName, this._port, this._dbName);
+  }
+}
+
+/* synchronization */
+
+function syncPeers(peers) {
+  peers.forEach(function (peer) {
+
+    var p = this._replications[peer.peerIdentifier];
+
+    if (!p && peer.peerAvailable) {
```

Comment:
You essentially repeat the same code to delete the client, replication and call disconnect both here and in syncRetry. I would suggest instead creating a syncPeer object which has on it a 'start' method that will create the client mux bridge, the replication and call connect. And then a stop method that will shut those down. And then you can have a retry which calls stop and then start. That will centralize the code, make it clearer what is involved in a peer replication and generally prevent bugs.